{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of the Human and Elephant P53 Genes\n",
    "\n",
    "## Jonathan Taylor"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this project, I analyze DNA sequences for the gene p53, which is used to suppress cancer in organisms. Specifically, I compare the p53 genes in humans and elephants, to identify what they have in common and how they are different.\n",
    "\n",
    "To do this, I interpret DNA data files from the NIH, convert them to RNA, then convert that to a sequence of generated proteins. Next, I compare the protein sequences of the two different genes, automatically generate text reports of your findings, and graph the results, which are displayed at the bottom of this notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data that is used for this project is retrieved from the NIH Gene Database.\n",
    "\n",
    "* Human p53 Gene:\n",
    "    - https://www.ncbi.nlm.nih.gov/nuccore/NC_000017.11?report=fasta&from=7668402&to=7687550&strand=true\n",
    "    \n",
    "* Elephant p53 Gene:\n",
    "    - https://www.ncbi.nlm.nih.gov/nuccore/NW_003573467.1?report=fasta&from=11687413&to=11699835&strand=true\n",
    "    \n",
    "\n",
    "DNA is composed of four bases: cytosine (C), guanine (G), adenine (A), and thymine (T). These bases, when read together, produce instructions that the organism can follow in order to create useful proteins. The connected sequence of bases is represented in a text file with a single letter per base."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 1: Generating Proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. Data Retrieval using readFile(filename) function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "from prettytable import PrettyTable\n",
    "project = \"ProteinSeq\" \n",
    "\n",
    "def readFile(filename):\n",
    "    d = open(filename, 'r')\n",
    "    text = d.read()\n",
    "    text = text.replace('\\n', '')\n",
    "    return text"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Transcribing DNA into RNA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that I have the data, the first thing I need to do is convert the DNA sequences into RNA sequences. I wrote the dnaToRna() function below to replicate the process of transcription. This process is not done by reading all of the DNA string. Instead, the organism breaks down the DNA into codons, or groups of three bases. The ATG codon signals the start of an RNA strand and three other codons (TAA, TAG, and TGA) signal the end (Stop). \n",
    "\n",
    "For this project, it is assumed that the starting location (startIndex) is known. So this function reads codons from that start point until the end point is found, and return a list of all the codons in between. In RNA, the T base is replaced with U."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dnaToRna(dna, startIndex):\n",
    "    \"\"\"\n",
    "    dnaToRna - converts DNA sequence into RNA sequence\n",
    "    \n",
    "    Arguments:\n",
    "        dna: string of DNA sequence that was loaded in Step 0 above\n",
    "        \n",
    "        startIndex: start codon of DNA sequence\n",
    "        \n",
    "    Returns:\n",
    "        rna: string of RNA sequence\n",
    "    \"\"\"\n",
    "    rna = []\n",
    "    aug = False\n",
    "    while startIndex + 2 < len(dna):\n",
    "        dna = dna.replace(\"T\", \"U\")\n",
    "        codon = dna[startIndex:startIndex+3]\n",
    "        if codon == \"AUG\" or aug == True:\n",
    "            aug = True\n",
    "            if codon == \"UAG\" or codon == \"UAA\" or codon == \"UGA\":\n",
    "                rna.append(codon)\n",
    "                return rna\n",
    "            else:\n",
    "                rna.append(codon)\n",
    "                dna = dna[3:]\n",
    "        else:\n",
    "            startIndex += 1\n",
    "    return rna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Building a Codon Dictionary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To turn RNA into proteins, I need to know which amino acid each codon corresponds to. This data is not formatted exactly as I need it to be; it maps amino acids to lists of codons, but I need to map each codon to a amino acid. This function returns a dictionary that maps each codon to its amino acid."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "def makeCodonDictionary():\n",
    "    \"\"\"\n",
    "    makeCodonDictionary - takes the \"codon_table.json\" file\n",
    "    (which maps amino acids to a list of codons) and maps \n",
    "    each codon to its amino acid\n",
    "    \n",
    "    Returns:\n",
    "        codonD: dictionary in which each codon is paired \n",
    "        with an amino acid\n",
    "    \"\"\"\n",
    "    codonD = {}\n",
    "    import json\n",
    "    c = open(\"codon_table.json\", \"r\")\n",
    "    aminoD = json.load(c)\n",
    "    for key in aminoD:\n",
    "        value = aminoD[key]\n",
    "        for codon in value:\n",
    "            codon = codon.replace(\"T\", \"U\")\n",
    "            codonD[codon] = key\n",
    "    return codonD"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Generating a List of Amino Acid Strings (Proteins)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To turn RNA into proteins, I need to identify each codon in the RNA sequence and add its associated amino acid to the chain. This chain of amino acids will become the protein. This function takes in the rna sequence generated from the rnaToDna() function and the codon dictionary generated from the makeCodonDictionary() function and returns a protein (a list of amino acid strings). To do this, I loop through each codon in the RNA list, and add its associated amino acid (based on the dictionary) to a new list. A special case is made for the first codon so that is says 'Start' instead of the amino acid name. \n",
    "\n",
    "\n",
    "Note: the proteins being generated here are not 100% accurate because this code skips over RNA splicing and other complex bioloical processes involved in gene expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generateProtein(codons, codonD):\n",
    "    \"\"\"\n",
    "    generateProtein - turns RNA sequence into \n",
    "    protein sequence\n",
    "    \n",
    "    Arguments:\n",
    "        codons: RNA sequence generated by dnaToRna \n",
    "        \n",
    "        codonD: dictionary generated by makeCodonDictionary\n",
    "        \n",
    "    Returns:\n",
    "        protein: list of amino acid strings\n",
    "    \"\"\"\n",
    "    protein = []\n",
    "    for i in range(len(codons)):\n",
    "        codon = codons[i]\n",
    "        if codon == \"AUG\":\n",
    "            if i == 0:\n",
    "                protein.append(\"Start\")\n",
    "            else:\n",
    "                protein.append(codonD[codon])\n",
    "        else:\n",
    "            a_acid = codonD[codon]\n",
    "            protein.append(a_acid)\n",
    "    return protein"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Synthesizing Proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This function calls upon all of the previously defined functions. \n",
    "\n",
    "* This function reads in the DNA by calling the readFile(filename) function and prints out the total number of bases in that DNA.\n",
    "\n",
    "* It produces a codon dictinary by calling the makeCodonDictionary() function.\n",
    "\n",
    "* It then identifies all of the RNA strands that can be produced from the DNA by iterating through all the indexes in the DNA string, looking for the start code (ATG) at each point.\n",
    "\n",
    "* If an index that corresponds to 'ATG' is identified, the dnaToRna(dna, startIndex) function is called from that index to extract the entire RNA sequence, then generateProtein(codon, codonD) is called to produce a protein.\n",
    "\n",
    "* The generated protein list is then added to another list of of proteins and the index in the DNA strand is updated to skip bases that were already checked.\n",
    "\n",
    "* If an index does not correspond to 'ATG', then index += 1 and this is an unused base.\n",
    "\n",
    "This function provides a list of all the proteins synthesized from the DNA and a count of all the bases that were not used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "def synthesizeProteins(filename):\n",
    "    \"\"\"\n",
    "    synthesizeProteins - ties all previous functions together\n",
    "    and creates a list of proteins from the original DNA sequence\n",
    "    \n",
    "    Argument:\n",
    "        filename: filename containing DNA sequences\n",
    "        \n",
    "    Returns:\n",
    "        protein_lst: list of all proteins synthesized from DNA\n",
    "    \"\"\"\n",
    "    dna = readFile(filename)\n",
    "    print(filename, \":\")\n",
    "    print(\"Length of DNA:\",len(dna))\n",
    "    codonDict = makeCodonDictionary()\n",
    "    protein_lst = []\n",
    "    unused_bases = 0\n",
    "    i = 0\n",
    "    while i < len(dna):\n",
    "        triplet = dna[i:i+3]\n",
    "        if triplet == \"ATG\":\n",
    "            rna_seq = dnaToRna(dna, i)\n",
    "            protein = generateProtein(rna_seq, codonDict)\n",
    "            protein_lst.append(protein)\n",
    "            i += 3 * len(rna_seq)\n",
    "        else:\n",
    "            i += 1\n",
    "            unused_bases += 1\n",
    "    print(\"number of proteins:\", len(protein_lst))\n",
    "    print(\"unused bases:\",unused_bases)\n",
    "    return protein_lst"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 2: Protein Sequence Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Finding Common Proteins Between Two Protein Lists"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This function begins the second part of this project, which is analyzing the protein sequences generated in the previous part. To determine how similar/different the genes are, I start by identifying any proteins that occur in both genes. The commonProteins function takes two protein lists generated from the synthesizeProteins function and returns a list of all the unique proteins in both lists."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "def commonProteins(proteinList1, proteinList2):\n",
    "    \"\"\"\n",
    "    commonProteins - shows proteins that occur in both genes\n",
    "    \n",
    "    Arguments:\n",
    "        proteinList1 and 2: two lists of proteins generated\n",
    "        by synthesizeProteins function\n",
    "        \n",
    "    Returns:\n",
    "        common_proteins: list of proteins that occur in \n",
    "        both genes\n",
    "    \"\"\"\n",
    "    common_proteins = []\n",
    "    for i in range(len(proteinList1)):\n",
    "        protein = proteinList1[i]\n",
    "        if protein in proteinList2 and protein not in common_proteins:\n",
    "            common_proteins.append(protein)\n",
    "    return common_proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Determining Amino Acid Frequencies"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It turns out that there are not very many proteins in common between the two genes, so a more interesting challenge is to compare the amino acids generated by the proteins to see if any amino acids occur more often in one gene over the other. \n",
    "\n",
    "To do this, first, I need to \"flatten\" the list of proteins into a list of the amino acids that occur. The combineProteins function takes a protein list and returns a list of all the amino acids that occur across all the proteins, in order.\n",
    "\n",
    "Once I have the list of amino acids, I use it to generate a dictionary that maps each amino acid in the list to a count of how often it occurs. This will help me answer the question: how common is each amino acid in the gene?\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "def combineProteins(proteinList):\n",
    "    \"\"\"\n",
    "    combineProteins - generates list of all the amino acids that \n",
    "    occur across all proteins, in order\n",
    "    \n",
    "    Argument:\n",
    "        proteinList: list of proteins generated by synthesizeProteins\n",
    "        \n",
    "    Returns:\n",
    "        combined_proteins: list of amino acids occuring in all proteins\n",
    "    \"\"\"\n",
    "    combined_proteins = []\n",
    "    for i in range(len(proteinList)):\n",
    "        for aa in range(len(proteinList[i])):\n",
    "            combined_proteins.append(proteinList[i][aa])\n",
    "    return combined_proteins\n",
    "\n",
    "\n",
    "def aminoAcidDictionary(aaList):\n",
    "    \"\"\"\n",
    "    aminoAcidDictionary - counts of each amino acid\n",
    "    \n",
    "    Argument:\n",
    "        aaList: list of amino acids generated by combineProteins\n",
    "        \n",
    "    Returns:\n",
    "        aa_dict: dictionary that maps each amino acid to a count of\n",
    "        how often it occurs\n",
    "    \"\"\"\n",
    "    aa_dict = {}\n",
    "    for i in range(len(aaList)):\n",
    "        key = aaList[i]\n",
    "        if key not in aa_dict:\n",
    "            aa_dict[key] = aaList.count(key)\n",
    "    return aa_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Sorting Amino Acid Frequencies"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To compare amino acid frequencies between genes of different lengths, I need my list of amino acids to be ordered by frequency: the rarest amino acids are at the front and the most common are at the back. If an amino acid occurs at significantly different places between two frequency lists, there is a major difference between the two organisms. The sortAminoAcidsByFreq function takes an amino acid list and returns a sorted frequency list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sortAminoAcidsByFreq(aaList):\n",
    "    \"\"\"\n",
    "    sortAminoAcidsByFreq - sorts list of amino acids by \n",
    "    frequency\n",
    "    \n",
    "    Argument:\n",
    "        aaList: amino acid list generated by combineProteins\n",
    "        \n",
    "    Returns:\n",
    "        freq_list: list of amino acids sorted from most frequent\n",
    "        to least frequent\n",
    "    \"\"\"\n",
    "    freq_list = []\n",
    "    aa_dict = aminoAcidDictionary(aaList)\n",
    "    for keys in aa_dict.keys():\n",
    "        frequency = []\n",
    "        occurance = aa_dict[keys]\n",
    "        prob = occurance / len(aaList)\n",
    "        frequency.append(prob)\n",
    "        if keys != \"Start\" and keys != \"Stop\":\n",
    "            frequency.append(keys)\n",
    "            freq_list.append(frequency)\n",
    "    freq_list.sort()\n",
    "    return freq_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8. Finding Differences Between Amino Acids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The findAminoAcidDifferences function takes two protein lists, one for each gene, identifies amino acids that occur at different indexes between the two lists and returns a list of those two amino acids and their frequencies. An amino acid has sufficiently different frequencies if it occurs at different indexes in the two frequency lists, and the difference between those two frequencies is greater than 0.5%."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "def findAminoAcidDifferences(proteinList1, proteinList2):\n",
    "    \"\"\"\n",
    "    findAminoAcidDifferences - finds differences\n",
    "    between genes\n",
    "    \n",
    "    Arguments:\n",
    "        proteinList1 and 2: two protein lists\n",
    "        (one for each gene) generated by \n",
    "        synthesizeProteins\n",
    "        \n",
    "    Returns:\n",
    "        comb_freq_list: list of amino acids that\n",
    "        occur at different indexes between the two \n",
    "        protein lists and their frequencies\n",
    "    \"\"\"\n",
    "    comb_freq_list = []\n",
    "    comb_prot1 = combineProteins(proteinList1)\n",
    "    comb_prot2 = combineProteins(proteinList2)\n",
    "    freq_list1 = sortAminoAcidsByFreq(comb_prot1)\n",
    "    freq_list2 = sortAminoAcidsByFreq(comb_prot2)\n",
    "    for i in range(len(freq_list1)):\n",
    "        for j in range(len(freq_list2)):\n",
    "            aa_list = []\n",
    "            aa1 = freq_list1[i][1]\n",
    "            aa2 = freq_list2[j][1]\n",
    "            if aa1 == aa2:\n",
    "                frequency1 = freq_list1[i][0]\n",
    "                frequency2 = freq_list2[j][0]\n",
    "                difference = abs(frequency1 - frequency2)\n",
    "                threshold = 0.005\n",
    "                aa_list.append(aa1)\n",
    "                aa_list.append(frequency1)\n",
    "                aa_list.append(frequency2)\n",
    "                if i != j and difference >= threshold:\n",
    "                    comb_freq_list.append(aa_list)\n",
    "    return comb_freq_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 9. Generating a Text Report of Amino Acid Differences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The displayTextResults function takes two lists: one list is the list of common proteins and the second list is the list of the most different amino acids between the two genes and prints out the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "def displayTextResults(commonalities, differences):\n",
    "    \"\"\"\n",
    "    displayTextResults - generates and displays \n",
    "    text report of similarities and differences between genes\n",
    "    \n",
    "    Arguments:\n",
    "        commonalities: list of common proteins between genes\n",
    "        \n",
    "        differences: list of the most different amino acids between genes\n",
    "    \"\"\"\n",
    "    print(\"The following proteins occurred in both DNA sequences:\")\n",
    "    for i in range(len(commonalities)):\n",
    "        if len(commonalities[i]) > 2:\n",
    "            for j in range(1, len(commonalities[i])-1):\n",
    "                print(commonalities[i][j], end=\" \")\n",
    "        print(\"\")\n",
    "\n",
    "    print(\"The following amino acids occurred at very different rates in the two DNA sequences (> 0.5% threshold):\")\n",
    "    for x in range(len(differences)):\n",
    "        for y in range(len(differences[x])):\n",
    "            aa = differences[x][0]\n",
    "            freq1 = (differences[x][1] * 100) \n",
    "            freq2 = (differences[x][2] * 100)\n",
    "        print(aa + \": \" + str(freq1) + \" % in Seq1, \" + str(freq2) + \" % in Seq2.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 3: Visualizing Results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 10: Generating Chart Labels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to plot our data using matplotlib, we first need to reformat it to fit the input that matplotlib expects. The makeAminoAcidLabels function takes a list of genes (where each gene is a protein list) and finds all the amino acids that occur across all the genes. This allows for comparison of more than two genes at a time and accounts for amino acids that may occur in one gene and not the other. It returns a sorted list of all the amino acids found."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "def makeAminoAcidLabels(geneList): \n",
    "    \"\"\"\n",
    "    makeAminoAcidLabels - reformats data for plotting\n",
    "    \n",
    "    Argument:\n",
    "        geneList: list of genes\n",
    "        \n",
    "    Returns:\n",
    "        xLabels: sorted list of all the amino acids found\n",
    "    \"\"\"\n",
    "    total_aa = []\n",
    "    sorted_aa_lst = [] \n",
    "    comb_prot = []\n",
    "    for i in range(len(geneList)):\n",
    "        comb_prot.append(combineProteins(geneList[i]))\n",
    "    for j in range(len(comb_prot)):\n",
    "        for k in range(len(comb_prot[j])):\n",
    "            total_aa.append(comb_prot[j][k])\n",
    "    comb_aa_dict = aminoAcidDictionary(total_aa)\n",
    "    xLabels = list(comb_aa_dict.keys())\n",
    "    xLabels.sort() #sorted aa_list\n",
    "    return xLabels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 11: Setting Up Chart Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The setupChartData function takes a labels list (produced by the makeAminoAcidLabels function) and a list of genes and returns a list of frequency lists such that each list's index i contains the frequency for the i'th element of the labels list. \n",
    "As an example, if Alanine (Ala) is in the 0th index of the labels list for the two p53 genes and I name my resulting list 'result', then result[0][0] is the frequency of Ala in the human gene and results[1][0] is the frequency of Ala in the elephant gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "def setupChartData(labels, geneList): \n",
    "    \"\"\"\n",
    "    seupChartData - generates a list of lists\n",
    "    for plotting with matplotlib\n",
    "    \n",
    "    Arguments:\n",
    "        labels: labels list generated by\n",
    "        makeAminoAcidLabels\n",
    "        \n",
    "        geneList: list of genes\n",
    "        \n",
    "    Returns:\n",
    "        freqList: list of frequency lists\n",
    "    \"\"\"\n",
    "    freqList = []\n",
    "    for i in range(len(geneList)):\n",
    "        c = combineProteins(geneList[i])\n",
    "        dict_aa = aminoAcidDictionary(c)\n",
    "        freq_lst = []\n",
    "        for j in range(len(labels)):\n",
    "            if labels[j] in dict_aa:\n",
    "                occurance = dict_aa[labels[j]]\n",
    "                prob = occurance / len(c)\n",
    "                freq_lst.append(prob)\n",
    "            else:\n",
    "                freq_lst.append(0)\n",
    "        freqList.append(freq_lst)\n",
    "    return freqList"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 12: Creating the Bar Chart and Edge Labels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The functions below generate a bar chart that illustrates the results of this analysis. The makeEdgeList function adds black edges around the amino acids that have significantly different frequencies between the two genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "def createChart(xLabels, freqList, freqLabels, edgeList=None):\n",
    "    \"\"\"\n",
    "    createChart - generates bar chart of results\n",
    "    \n",
    "    Arguments:\n",
    "    xLabels: labels list generated by makeAminAcidLabels\n",
    "    \n",
    "    freqList: list of frequency lists generated by setupChartData\n",
    "    \n",
    "    freqLabels: list of strings, where each string is the name of \n",
    "    the gene represented by the corresponding list in freqLists\n",
    "    \n",
    "    edgeList\n",
    "    \"\"\"\n",
    "    import matplotlib.pyplot as plt\n",
    "    import numpy\n",
    "    \n",
    "    x = numpy.arange(len(xLabels))\n",
    "    width = 0.6 / len(freqList)\n",
    "    fig, ax = plt.subplots()\n",
    "    for i in range(len(freqList)):\n",
    "        offset = -0.3 + width/2\n",
    "        ax.bar(x + offset + i*width, freqList[i], width, edgecolor=edgeList)\n",
    "    \n",
    "    ax.set_ylabel(\"Frequencies\")\n",
    "    ax.set_title(\"Protein Sequencing\")\n",
    "    ax.set_xticks(x)\n",
    "    ax.set_xticklabels(xLabels, Fontsize=7)\n",
    "    ax.legend([\"Human p53\", \"Elephant p53\"])\n",
    "    \n",
    "    fig.tight_layout()\n",
    "    plt.show()\n",
    "\n",
    "def makeEdgeList(labels, biggestDiffs):\n",
    "    \"\"\"\n",
    "    makeEdgeList - adds black edges around the amino acids with \n",
    "    significantly different frequencies between the two genes\n",
    "    \n",
    "    Arguments:\n",
    "        labels: labels list generated by makeAminoAcidLabels\n",
    "        \n",
    "        biggestDiffs: list of most different amino acids\n",
    "        generated by findAminoAcidDifferences\n",
    "        \n",
    "    Returns:\n",
    "        edge_lst: list, the same length as the labels list, where each \n",
    "        element of the list is \"black\" if the corresponding amino acid \n",
    "        is in the biggestDiffs list, and \"white\" otherwise.\n",
    "    \"\"\"\n",
    "    lst = []\n",
    "    edge_lst = []\n",
    "    for j in range(len(biggestDiffs)):\n",
    "        lst.append(biggestDiffs[j][0])\n",
    "    for i in range(len(labels)):\n",
    "        if labels[i] in lst:\n",
    "            edge_lst.append(\"black\")\n",
    "        else:\n",
    "            edge_lst.append(\"white\")\n",
    "    return edge_lst"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 13: Putting Everything Together"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The runFullProgram function brings all the functions defined and discussed above together for the final result. It loads the DNA data in the two p53 files, processes them both into protein lists, generates a text report comparing the two genes, and then generates a bar chart comparing the two genes with the sufficiently different amino acids outlined in black."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "human_p53.txt :\n",
      "Length of DNA: 19149\n",
      "number of proteins: 119\n",
      "unused bases: 10560\n",
      "elephant_p53.txt :\n",
      "Length of DNA: 12423\n",
      "number of proteins: 77\n",
      "unused bases: 6204\n",
      "The following proteins occurred in both DNA sequences:\n",
      "Ser Pro Leu \n",
      "Lys \n",
      "Ala \n",
      "Thr \n",
      "Gly \n",
      "The following amino acids occurred at very different rates in the two DNA sequences (> 0.5% threshold):\n",
      "Tyr: 2.060775410408662 % in Seq1, 2.6049204052098407 % in Seq2.\n",
      "Met: 2.1306322039818375 % in Seq1, 3.087313072841293 % in Seq2.\n",
      "Ile: 3.0736989172196996 % in Seq1, 2.460202604920405 % in Seq2.\n",
      "Phe: 3.9818372336709746 % in Seq1, 5.258080077182827 % in Seq2.\n",
      "Thr: 4.016765630457562 % in Seq1, 4.82392667631452 % in Seq2.\n",
      "Lys: 4.959832343695425 % in Seq1, 4.196816208393632 % in Seq2.\n",
      "Arg: 6.007684247293049 % in Seq1, 4.630969609261939 % in Seq2.\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAagAAAEYCAYAAAAJeGK1AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAneElEQVR4nO3dfZxVZbn/8c+XYWRQFAo1kREGy/IBgnAENE/SzzTxIfSkptmDmngsqTD7JSfPSexYkZmpxxS1zIcs7WAaIqVZmj8yBFRSwVRUiEHooAk+MTLA9ftjrRk32z0ze2DWzJqZ7/v1mhd7PVx7XXvYs6993+te91JEYGZmlje9OjsBMzOzUlygzMwsl1ygzMwsl1ygzMwsl1ygzMwsl1ygzMwsl1ygzLaBpMWSxnd2Hp1N0imS7u3sPKx7ka+Dsq5O0jLgPcAm4A1gDvDliHh9K54rgL0iYmm7Jln6WBOBC4E9gQ3AX4EvRMSyrI9t1hW4BWXdxTER0Q8YDRwA/EfxDpJ6d3hWzZD0PuAm4FygPzAMuArY3Jl5meWJC5R1KxGxEvgtMBySFpGksyU9Czybrpskaamkf0qaJWn3dP2D6dP8VdLrkj6Vrj9a0iJJayU9JOmDjceTtEzSx9LH0yT9StJNkl5Lu/9qm0l1FPBCRPwhEq9FxO0R8ff0uXpJmirpOUkvp8/77oLjflbS8nTb+UV53CDpooJ9x0uqK1jeXdLtktZIekHSVwq2tfgaJO0h6ddp7MuSrkzXnyppbsF+IeksSc9KekXSjyUp3VYh6YeSXkqPPzndPzdfICwfXKCsW5G0B3Ak8FjB6mOBscC+kv4P8D3gRGAQsBy4FSAiPpLuPzIi+kXEbZJGA9cD/wYMBK4BZknq00wKn0ifbwAwC7iymf0eBfaW9CNJH5XUr2j7V9K8DwF2B14Bfpy+xn2Bq4HPptsGAtXN/lIKSOoF3EXSnTgYOBSYIunjrb0GSRXAbJLfWU0af2sLhzuapDU7kuT33XiMScAEkiI9On2dZu/gAmXdxZ2S1gJzgT8B3y3Y9r2I+GdErAdOAa6PiEcj4i3g34EDJdU087yTgGsi4uGI2BQRNwJvAeOa2X9uRMyJiE3AzSQfzu8QEc8D40k+5H8FvJS2fBoL1b8B50dEXZrnNOD4tJVxPDA7Ih5Mt/0n5XcNHgDsEhHfjogNaR7XASeV8RrGkBTE/xsRb0REfUTMpXnTI2Jt2iq8n6QgQVKsLk9f2yvA9DJztx7GTWrrLo6NiPua2bai4PHuJK0XACLidUkvkxSKZSVihwKfl/TlgnXbpc9TyuqCx28CVZJ6R8TG4h0jYh7JhzWSDgBuA84nKZpDgTskFRaeTSSDQXYvfE0R8Ub6GsoxFNg9LeaNKoD/19prAPYAlpd6Lc0ofp7G4rtF/kWPzZq4QFlPUDhU9UWSD2kAJO1A0kW2spnYFcB3IuI72aUHEbFA0q9Jz52lxz09Iv5cvK+kVcA+Bcvbk7yGRm8A2xcs71bweAXJua+9tiLNFcCQ5gpuG6xiyy7JPbbhuawbcxef9TS/AE6TNCo9j/Rd4OGCod3/IBn23eg64CxJY5XYQdJRknbcliQkHZwO1tg1Xd6b5NzPvHSXGcB3JA1Nt++SDksHmAkcnT7HdsC32fJveRFwpKR3S9oNmFKwbT7wqqTzJPVNBywMT1twrZlPUlymp7+HKkkf3oqX/yvgq5IGSxoAnLcVz2E9gAuU9SgR8QeScza3k3zYvpctz79MA25MR+ydGBELSc5DXUkyUGEpcGo7pLKWpCA9Iel14HfAHcDF6fbLSQYo3CvpNZLCNTZ9DYuBs0mK7ao0r7qC576ZZBDEMuBekq5D0thNwDGkowiBl4CfkAx1b1FB7PuAv6fH/FTbXjaQFP17gcdJBrPMATaSdGGaNfGFumbdgJKLlc9o4TxcbkmaAMyIiKGt7mw9iltQZtah0q7FIyX1ljQYuICk9Wi2BRcoM+toIpni6RWSLr6ngG91akaWS+7iMzOzXHILyszMcqlbXQe18847R01NTWenYWZmbfDII4+8FBG7FK/vVgWqpqaGhQsXdnYaZmbWBpKWl1rvLj4zM8slFygzM8slFygzM8ulbnUOyswMoKGhgbq6Ourr6zs7FStQVVVFdXU1lZWVZe3vAmVm3U5dXR077rgjNTU1pDfytU4WEbz88svU1dUxbNiwsmLcxWdm3U59fT0DBw50ccoRSQwcOLBNrVoXKDPrllyc8qet/ycuUGZmlksuUGbW7dU3tO+tpsp5vn79+m2xfMMNNzB58uR2zaO9TZs2jcGDBzNq1ChGjRrFnDlzAJg/f37TupEjR3LHHR0z+bwHSZhZt1dVWUHN1Lvb7fmWTT+q3Z4rb8455xy+/vWvb7Fu+PDhLFy4kN69e7Nq1SpGjhzJMcccQ+/e2ZYQt6DM8qChxInjUuusWzj11FOZOXNm03Jja+uBBx7gkEMO4cQTT+T9738/U6dO5ZZbbmHMmDGMGDGC5557DoC77rqLsWPH8qEPfYiPfexj/OMf/wCSFtDpp5/O+PHj2XPPPbniiitKHr9fv36ce+65jB49mkMPPZQ1a9a0mO/222/fVIzq6+s77PyeC5RZHlRWwbT+W/5UVnV2VrYN1q9f39QtNmrUKL71rfJuefXXv/6Vyy+/nCeeeIKbb76ZZ555hvnz53PGGWfw3//93wAcfPDBzJs3j8cee4yTTjqJiy++uCn+b3/7G/fccw/z58/nwgsvpKGh4R3HeOONNxg9ejSPPvoohxxyCBdeeGHTtiuvvJIPfvCDnH766bzyyitN6x9++GH2228/RowYwYwZMzJvPYELlJlZJvr27cuiRYuafr797W+XFXfAAQcwaNAg+vTpw3vf+14OP/xwAEaMGMGyZcuA5Dqvj3/844wYMYIf/OAHLF68uCn+qKOOok+fPuy8887suuuuTa2rQr169eJTn/oUAJ/5zGeYO3cuAF/84hd57rnnWLRoEYMGDeLcc89tihk7diyLFy9mwYIFfO973+uQi6BdoMzMOljv3r3ZvHkzkFzAumHDhqZtffr0aXrcq1evpuVevXqxceNGAL785S8zefJknnjiCa655potikVhfEVFRVNMSxq77N7znvdQUVFBr169mDRpEvPnz3/Hvvvssw877LADTz75ZFte8lZxgTIz62A1NTU88sgjAPzmN78p2Q3XknXr1jF48GAAbrzxxjYff/PmzU3nwH7xi19w8MEHA7Bq1aqmfe644w6GDx8OwAsvvNBU6JYvX87TTz9NR9x7z6P4zKzbq2/Y1K4j7+obNlFVWbHV8ZMmTWLixImMGTOGQw89lB122KFN8dOmTeOEE05g8ODBjBs3jhdeeKFN8TvssAOLFy9m//33p3///tx2220AfOMb32DRokVIoqamhmuuuQaAuXPnMn36dCorK+nVqxdXXXUVO++8c5uOuTUUEZkfpKPU1taGb1hoXda0/kXL6zonj27gqaeeYp999unsNHKrX79+vP76651y7FL/N5IeiYja4n3dxWdmZrnkAmVm1sN0VuuprVygzMwsl1ygzDpBe88NZ9YdeRSfWSconhuuO8/tZra13IIyM7NccoEys+6vvSfeLeP5KioqtpiLb/r06QCMHz+erb0cpniS2W1xww038OKLL7Yp5oEHHqB///5Nr6lx+qb6+nrGjBnDyJEj2W+//bjgggvaJUd38ZlZ99c4GW97KeMatca5+PLqhhtuYPjw4ey+++5tivuXf/kXZs+evcW6Pn368Mc//pF+/frR0NDAwQcfzIQJExg3btw25egWlJlZJ7n33ns58MADGT16NCeccELT8O+amhrOO+88xowZw5gxY1i6dGlTzIMPPshBBx3Ennvu2dSaev311zn00EMZPXo0I0aM4De/+Q0Ay5YtY5999mHSpEnst99+HH744axfv56ZM2eycOFCTjnlFEaNGsX69eu3yGv8+PFMmTKFgw46iOHDh5eck6+QpKZbhjQ0NNDQ0NAut+RwgTIzy0Dx7TYapxNq9NJLL3HRRRdx33338eijj1JbW8ull17atH2nnXZi/vz5TJ48mSlTpjStX7VqFXPnzmX27NlMnToVgKqqKu644w4effRR7r//fs4991waZwl69tlnOfvss1m8eDEDBgzg9ttv5/jjj6e2tpZbbrmFRYsW0bdv33fk/8Ybb/DQQw9x1VVXcfrppzet/8tf/sLIkSOZMGHCFrOob9q0iVGjRrHrrrty2GGHMXbs2G3+HbqLz8wsA6118c2bN48lS5bw4Q9/GIANGzZw4IEHNm0/+eSTm/4955xzmtYfe+yx9OrVi3333bfpVhoRwTe/+U0efPBBevXqxcqVK5u2DRs2jFGjRgGw//77N92yozWNx//IRz7Cq6++ytq1axk9ejTLly+nX79+zJkzh2OPPZZnn30WSM65LVq0iLVr13Lcccfx5JNPNk02u7VcoMzMOkFEcNhhh/HLX/6y5PbCLrLCx4W302hsJd1yyy2sWbOGRx55hMrKSmpqappuwVF8+43i7rzmFHfRSWKnnXZqWj7yyCP50pe+xEsvvbTFxLEDBgxg/Pjx/O53v9vmAuUuPjOzTjBu3Dj+/Oc/N51fevPNN3nmmWeatjd2Cd52221btKxKWbduHbvuuiuVlZXcf//9LF++vNXj77jjjrz22mvNbm88/ty5c+nfvz/9+/dn9erVTUVx/vz5bN68mYEDB7JmzRrWrl0LJF2b9913H3vvvXerObQm0xaUpCOAy4EK4CcRMb1o+97Az4DRwPkRcUm5sWZmZWuob9/Z4Rvqk5GBLWg8B9XoiCOOaBpqDrDLLrtwww03cPLJJ/PWW28BcNFFF/H+978fgLfeeouxY8eyefPmZltZjU455RSOOeYYamtrGTVqVFnF4dRTT+Wss86ib9++/OUvf3nHeah3vetdHHTQQbz66qtcf/31AMycOZOrr76a3r1707dvX2699VYksWrVKj7/+c+zadMmNm/ezIknnsjRRx/dag6tyex2G5IqgGeAw4A6YAFwckQsKdhnV2AocCzwSmOBKie2FN9uw7qSd8wk4dtttJuufruNmpoaFi5c2CH3XCpl/PjxXHLJJdTWvuMOGNssL7fbGAMsjYjnI2IDcCswsXCHiPjfiFgAFN9OstVYMzPr3rLs4hsMrChYrgPKHXdYdqykM4EzAYYMGdL2LM3McqbckXZZeeCBBzr1+I2ybEGVukqr3P7EsmMj4tqIqI2I2l122aXs5Myse+tOdwvvLtr6f5JlgaoD9ihYrgbKnfhpW2LNrIerqqri5ZdfdpHKkYjg5Zdfpqqq5cElhbLs4lsA7CVpGLASOAn4dAfEmlkPV11dTV1dHWvWrOnsVKxAVVUV1dXVZe+fWYGKiI2SJgP3kAwVvz4iFks6K90+Q9JuwEJgJ2CzpCnAvhHxaqnYrHI1s+6lsrKSYcOGdXYato0yvQ4qIuYAc4rWzSh4vJqk+66sWDMz6zk8k4SZmeWSC5SZmeWSC5SZmeWSC5SZmeWSC5SZmeWSC5SZmeWSC5SZmeWSC5SZmeWSC5SZmeWSC5SZmeWSC5SZmeWSC5SZmeWSC5SZmeWSC5SZmeWSC5SZdV0N9S0vW5eW6f2gzMwyVVkF0/q/vTxtXeflYu3OLSgzM8slFygzM8slFygzM8slFygzM8slFygz6xLqGzZ1dgrWwTyKz8y6hKrKCmqm3r3FumXTj+qkbKwjuAVlZma55AJlZma55AJlZma55AJlZma55AJlZma55AJlZma55AJlZma55AJlZma55AJlZma5lGmBknSEpKclLZU0tcR2Sboi3f64pNEF286RtFjSk5J+Kakqy1zNzCxfMitQkiqAHwMTgH2BkyXtW7TbBGCv9OdM4Oo0djDwFaA2IoYDFcBJWeVqZmb5k2ULagywNCKej4gNwK3AxKJ9JgI3RWIeMEDSoHRbb6CvpN7A9sCLGeZqZmY5k2WBGgysKFiuS9e1uk9ErAQuAf4OrALWRcS9pQ4i6UxJCyUtXLNmTbslb2ZmnSvLAqUS66KcfSS9i6R1NQzYHdhB0mdKHSQiro2I2oio3WWXXbYpYTMzy48sC1QdsEfBcjXv7KZrbp+PAS9ExJqIaAB+DRyUYa5mZpYzWRaoBcBekoZJ2o5kkMOson1mAZ9LR/ONI+nKW0XStTdO0vaSBBwKPJVhrmZmljOZ3bAwIjZKmgzcQzIK7/qIWCzprHT7DGAOcCSwFHgTOC3d9rCkmcCjwEbgMeDarHI1M7P8yfSOuhExh6QIFa6bUfA4gLObib0AuCDL/MzMLL88k4SZmeWSC5SZmeWSC5SZmeWSC5SZdZ6G+vLWWY+U6SAJM7MWVVbBtP5brpu2rnNysdxxC8rMOkx9w6bOTsG6ELegzKzDVFVWUDP17qblZdOP6sRsLO/cgrIuq9S3cX9DN+s+3IKyLqv42zj4G7lZd+IWlJmZ5ZILlJmZ5VJZBUrSeyX1SR+Pl/QVSQMyzczMzHq0cltQtwObJL0P+CnJjQR/kVlWZmbW45VboDZHxEbgOOCyiDgHGJRdWmZm1tOVW6AaJJ0MfB6Yna6rzCYlMzOz8gvUacCBwHci4gVJw4CfZ5eWmZn1dGVdBxURSySdBwxJl18ApmeZmJmZ9WzljuI7BlgE/C5dHiVpVoZ5mZlZD1duF980YAywFiAiFpGM5DMzM8tEuQVqY0QUz4Ef7Z2MmZlZo3Ln4ntS0qeBCkl7AV8BHsouLTMz6+nKbUF9GdgPeAv4JfAqMCWjnMzMzMoexfcmcH76Y5ZfDfXJXVpbW2dmuddigZJ0WURMkXQXJc45RcQnMsvMbGv4FuJm3UZrLaib038vyToRMzOzQi0WqIh4JH24EFgfEZsBJFUAfTLOzczMerByB0n8Adi+YLkvcF/7p2NmZpYot0BVRcTrjQvp4+1b2N/MzGyblFug3pA0unFB0v7A+mxSMjMzK/9C3SnA/0h6MV0eBHwqk4zMzMwo/zqoBZL2Bj4ACPhbRDRkmpmZmfVo5bagAA4AatKYD0kiIm7KJCszM+vxyr3dxs0k10IdTFKoDgBqy4g7QtLTkpZKmlpiuyRdkW5/vOg81wBJMyX9TdJTkg4s+1WZmVmXV24LqhbYNyLKnsE8vVbqx8BhQB2wQNKsiFhSsNsEYK/0ZyxwdfovwOXA7yLieEnb4VGDZmY9Srmj+J4Edmvjc48BlkbE8xGxAbgVmFi0z0TgpkjMAwZIGiRpJ+AjwE8BImJDRKxt4/HNzKwLK7dA7QwskXSPpFmNP63EDAZWFCzXpevK2WdPYA3wM0mPSfqJpB1KHUTSmZIWSlq4Zs2aMl9ONgZVD0FS08+g6iGdmo+ZWVdWbhfftK14bpVYV9xF2Nw+vYHRwJcj4mFJlwNTgf98x84R1wLXAtTW1nbqTRRXr1zB0PNmNy0v//7RnZiNmVnXVu4w8z9JGgrsFRH3SdoeqGglrA7Yo2C5GnixzH0CqIuIh9P1M0kKlJmZ9RDljuKbRFIkrklXDQbubCVsAbCXpGHpIIeTgOJuwVnA59LRfOOAdRGxKiJWAyskfSDd71BgCWZm1mOU28V3Nsmgh4cBIuJZSbu2FBARGyVNBu4haW1dHxGLJZ2Vbp8BzAGOBJYCbwKnFTzFl4Fb0uL2fNG2TNQ3bKKqsqLVdWZmlr1yC9RbEbFBSk4ZSepNiRsYFouIOSRFqHDdjILHQVL8SsUuooxrrdpTVWUFNVPv3mLdsulHdWQKZmaWKncU358kfRPoK+kw4H+Au7JLy8zMerpyC9RUkmHfTwD/RtIq+o+skjIzMyt3FN9m4Lr0x8zMLHNlFShJL1DinFNE7NnuGZmZmdG2ufgaVQEnAO9u/3RyqKEeKqtaX2dmZu2q3C6+l4tWXSZpLvCt9k8pZyqrYFr/LddNW9c5uZiZ9SDldvGNLljsRdKi2jGTjMzMzCi/i++HBY83AsuAE9s9GzMzs1S5XXwfzToRMzOzQuV28X2tpe0RcWn7pGNmXU1N9SCWr1zdtDx08G4sq1vViRlZd9GWUXwH8PZkr8cAD7LlvZzMrAdavnI1ccFOTcu6cHULe5uVr9wCtTMwOiJeA5A0DfifiDgjq8TMzKxnK3eqoyHAhoLlDUBNu2djZmaWKrcFdTMwX9IdJDNKHAfclFlWZmbW45U7iu87kn4L/Eu66rSIeCy7tMzMrKcrt4sPYHvg1Yi4HKiTNCyjnMysBYOqhyBpi59B1UM6Oy2zdlfuMPMLSEbyfQD4GVAJ/Bz4cHapmVkpq1euYOh5s7dYt/z7R3dSNmbZKbcFdRzwCeANgIh4EU91ZGZmGSq3QG1Ib88eAJJ2yC4lM7PuraZ60Du6aWuqB3V2WrlT7ii+X0m6BhggaRJwOr55oZnZVim+uBnyeYFzfcMmqiorml3OWqsFSpKA24C9gVdJzkN9KyJ+n3FuZmbWiaoqK6iZenfT8rLpR3Xo8VstUBERku6MiP0BFyUzM+sQ5Z6DmifpgEwz6Yb6VOB+ZjOzrVTuOaiPAmdJWkYykk8kjasPZpVYd/DWJrpEP7OZWR61WKAkDYmIvwMTOigfMzMzoPUW1J0ks5gvl3R7RHyyA3IyMzNr9RyUCh7vmWUiZmZmhVorUNHMY7N2Ud+wqax1ZpYDDfXlrWsnrXXxjZT0KklLqm/6GN4eJLFT86FmrSu+zgI6/loLMytTZRVM67/lumnrMjtciwUqIjrukmEzM7MCbbndhpmZWYfJtEBJOkLS05KWSppaYrskXZFuf1zS6KLtFZIekzS7ONa6seI+7Qz7uLuL4ovCfUH4lrb1XGfxvj5P2jHKvVC3zSRVAD8GDgPqgAWSZkXEkoLdJgB7pT9jgavTfxt9FXgK8LmunqS4nzvDPu7uoviicF8QvqVtPdfZ2XPS9VRZtqDGAEsj4vmI2ADcCkws2mcicFMk5pHMlj4IQFI1cBTwkwxzNDOznMqyQA0GVhQs16Xryt3nMuAbwOaM8jMzy9yg6iHvmJPTypNZFx9bXuTbqPhaqpL7SDoa+N+IeETS+BYPIp0JnAkwZMiQrUjTzMo1qHoIq1euaH1Ha7J65QqGnrflafTl3z+6k7LpWrIsUHXAHgXL1cCLZe5zPPAJSUcCVcBOkn4eEZ8pPkhEXAtcC1BbW+uLic0y5A9b60hZdvEtAPaSNEzSdsBJwKyifWYBn0tH840D1kXEqoj494iojoiaNO6PpYqTmZl1X5m1oCJio6TJwD1ABXB9RCyWdFa6fQYwBzgSWAq8CZyWVT5mZta1ZNnFR0TMISlChetmFDwO4OxWnuMB4IEM0jMzsxzzTBK2TTzZq3VZ2zLxaQdPmtpTZdqCsu6v5AWQ/3UoSa9ugYb65AJcs7zYlolPO3jS1J7KBcpKKy4obSkw/uM1s3bgAmWlebohM+tkPgdlZma55AJlZma55AJlHnVnZrnkc1Dm266bWS65BWVmZrnkAmVmZrnkAmVmZrnkAmVmZrnkAmXWyPOrmeWKC5RZo8bZMwp/mpneqdRtvAdVd807OtdUD3rHa6mpHtTZaZl5mLnZ1uhOd5ZdvnI1ccFOW6zThas7KRuzt7kFZWZmueQCZT2WZ9Awyzd38VmPVTyDhmfPMMsXt6DMzCyXXKDMzCyXXKDMzCyXXKDMzCyXXKDMzCyXXKDMzHqQ4plD8jxriIeZm5n1IMUzh+R51hC3oPKseKJST1xqZj2IW1B51jh5aaNp6zovF2tVnwqQtMW6oYN3Y1ndqk7KyKxrc4EyaydvbcKTrpq1I3fxmZlZLrlAmZlZLrlAmZlZLrlAmZlZLmVaoCQdIelpSUslTS2xXZKuSLc/Lml0un4PSfdLekrSYklfzTJPM7O8Kr5vWU+6j1lmo/gkVQA/Bg4D6oAFkmZFxJKC3SYAe6U/Y4Gr0383AudGxKOSdgQekfT7olgz2wqDqoeweuWKzk7DytST71uW5TDzMcDSiHgeQNKtwESgsMhMBG6KiADmSRogaVBErAJWAUTEa5KeAgYXxZrZVli9cgVDz5vdtLz8+0d3YjbWZg31yTWSra2j638ZybJADQYKfzN1JK2j1vYZTFqcACTVAB8CHi51EElnAmcCDBkyZFtz7jT1DZuoqqxovydsw5vYzLqQ4gv4odmL+Iu/jEDX+kKSZYFSiXXRln0k9QNuB6ZExKulDhIR1wLXAtTW1hY/f5dR3IyHbWzKt+FNbGaWR1kOkqgD9ihYrgZeLHcfSZUkxemWiPh1hnmamVkOZVmgFgB7SRomaTvgJGBW0T6zgM+lo/nGAesiYpWSCc1+CjwVEZdmmGO30ZNG9phZz5BZF19EbJQ0GbgHqACuj4jFks5Kt88A5gBHAkuBN4HT0vAPA58FnpC0KF33zYiYk1W+XV1PHuljZt1TppPFpgVlTtG6GQWPAzi7RNxcSp+fMjOzHsIzSZiZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS65QJmZWS5lWqAkHSHpaUlLJU0tsV2Srki3Py5pdLmxZmbWvWVWoCRVAD8GJgD7AidL2rdotwnAXunPmcDVbYg1M7NuLMsW1BhgaUQ8HxEbgFuBiUX7TARuisQ8YICkQWXGmplZN6aIyOaJpeOBIyLijHT5s8DYiJhcsM9sYHpEzE2X/wCcB9S0FlvwHGeStL4APgA83c4vZWfgpU6I9bG7VqyP3bOO3VXz7uxjN2doROxSvLJ3BgdqpBLriqthc/uUE5usjLgWuLZtqZVP0sKIqO3oWB+744/dVfP2sf1e6SrHbqssC1QdsEfBcjXwYpn7bFdGrJmZdWNZnoNaAOwlaZik7YCTgFlF+8wCPpeO5hsHrIuIVWXGmplZN5ZZCyoiNkqaDNwDVADXR8RiSWel22cAc4AjgaXAm8BpLcVmlWsrtqX7cFu7Hn3srhPrY/esY3fVvDv72G2S2SAJMzOzbeGZJMzMLJdcoMzMLJeyHMXX5Uj6EHAnsDfJrBaTI+L1tsZGxPp0Xa+I2Lw1sVubd1vj00EoPyQZ2l8J/CIi/tTG+EtIvuwIeAQ4MiKOzzK2neJPBV6KiNmS9ga+BjwVET9qQ3w/YFlEzC4npvi45ca0R3waNyUiRkmqJDn3e3bh80iqIXnff72Z+OOBl4HFwJsRcWUbj98UHxEXt+FvZD/gQmA18PeIuLiZ/ZqerzgGGNPae0PSMJLfyde3Jr7E8/UFfgTsDrwLeANYHxHHtZR7C/GLgR8Az0cZ52eai4+I54r2U/HzSfo18OmIqJc0AjglIqYWbJ/Z1t9HW7lAbelU4Fskf0QASDoKOATYFfhaRPyztVhJhwLLgCck1ZJ8eH4A+GkzHyqFsfsCVcCKiLhU0hPADcD+wBeaKUAtxS8BrgFGAF+NiDeKYicBcyLit+nr/TXwp3RexMOBBmAoyQjL/yxx7EnAbwvie5MMfKGM3FuKnRkRx6eDav4WEQ+Ue+z0d/7RiPiBpB8Bl0bEihLxxbYDBkvaB7iA5EPppoh4srVASUOBc0n+r5+LiMvKOB6SDgOOAvoCt5N8kDQWzVsj4qQWYo8HekXEryTdSPIe+AHJe292RDxYIuxpSQcB7wH+AgyUdEVj3iQXYR4kaXIzxWdGmtvLwJ/TD/Q+ETFZ0qeAA4GdgCsiYlFL8ZI+CCyU9AJwDMn7dlpELC0RdzjJ30/j//X5wC7AjsAU4JfAQyQjgH9fHCPpYOBUSf8BXA+cDQwEfhcRd0paDNxM8h44KB2kVdlM/NXAZcCrwOMRcU1zf2fpe/4sSeOB4cBs4EZJF5NcSvNp4Geknxck74EmJeJPBk4AnpP0eeAPwK4R8e8lfmfF8QcD90bEc5J+RTIpws+Bu4CfAmuKwm8HPgncApwOvEfSlcBfI+K6Usdrb+7iS0mqAt5N8p9xTMGmTSS/p0rgY22IvQ64D3hP+q3joTJjd0n3/Um6S11E/JDkw2TUVsS/GBGXA3cDnyiRwn4kf9SNZkv6GHBG+hy7AQuBK0rl3xgvqZeky4DCD7UWc28lthwl4yNiIfBBSQOAnVopTl+SNAOYVrCuP/AKSWuy1eLU+DzAepLWwYg2vIavAGuBVSRTfLXFHcAxaZFYBWwmeb/eCfy/ZmJmknzofBy4Nz1+Yd5zgYdaaBlNknQdyYjbhyPiXJL3HMDk9LX8o4XX0hh/IcmXi8uAz6WzxpwPnNVM3E+BgyX9VNJXgY+kx9oA7EPyN3pxRPy+VAzJe2VJRFwEbAT6pHmeku77YkRMB24qeP3NxZ9MMrL4bN7+TGjt76zQMxHxjfT4u6XrrouI21uIabQkzXMNMC9t7W+vZIq41rwGnCbpEOBPJJMfLImI6RFRXJwgKVDHpT0VQ0i+sP0TOLGMY7ULt6De9klgEMmH3AdImsIAX4yIiem3le3bELuOZIh8Y7O5ueZ4ceyJ6fKtJK2JxhZPA8kfVVvjG/+PK5vJYTFJC+eedPkWkm+SayLiJUnnAQcAP5P06Yh4tUT8Aek32ymSZhZsay33lmIbuzp2KBFXTvyvSX4H320hHuCqgi6+qQARMU/SSuALkkZGxI2tPAckH5A3R8TjZexbHHdRRGwEkHQyb/+ftfTaiYhNkh4l+b+fHBErJH0F+FfgMJIiUKyxFbua5HfcC7ilMW9JQ1rJ97r09zWe5D0Ob7+v1kfEtDLjTy2Ib3pJzQWl77vz0xzvBx4rPJakNxt/h83E3E1yKQvAZ0muq3wY+E26rjGXzWXEq0Surf2dFWo8Vj1v/10U/y5aiy0+Zjk2Aa8D55D8Dga2dNy0a29Vuv+sZFXclP7+O4QL1Ns+CRwTEevTrofz0/VL0u6EfUhaRGXHRsQ6Sf8raTrJN7CFZcSeQfIH8PxW5l0cP1DSd4Fh6bZi1wGXSjqGpKDeSvItqbEF9g2S+bf+ydt/oMXxl0g6muSb6UKg3KlQWopdKenrJN0Sj2xF/F3A15vp5mqRpI+StEbfzdvdRa25Evhu+gf9WkSUKg6FvpTm/SDwE0n/TPN/ALg4bRUNKCP+RWBjRLyQnif4AklXWXPvVUj+TwP4HElX1XmNeQPfA94n6WsRcWkrr6HYzyVdS1IE746Ie9sQdw3JF8D/KrWDpGNJWn2N/8+SdClJ12jJLyFFMUtIum//i+T6y7OAD5O0wAqtJn39JH9DpeIvB34k6V+Bxg/r1v7OsjBG0vdJvhisKjPmduDkiHhN0sAy9v8ZSUv7EOCCtKVWsXXptp2vg8pY+iY+gKQb5EuRzM7ekcdv04nMtKV4QJSYmLerSLs9ryA5t3ZnJ6eTmfTD4gck53vmd3Y+PVlb/87a4XjjgeEtdMWWitmX5NzqNyJieUaptSsXKDMzyyUPkjAzs1xygTIzs1xygTIzs1xygTIzs1xygTIzs1z6/zgAGKdcSN9eAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "def runFullProgram():\n",
    "    \"\"\"\n",
    "    Pulls all functions defined in this program together to process, \n",
    "    analyze and display similarities and differences between the \n",
    "    human and elephant p53 genes\n",
    "    \"\"\"\n",
    "    humanProteins = synthesizeProteins(\"human_p53.txt\")\n",
    "    elephantProteins = synthesizeProteins(\"elephant_p53.txt\")\n",
    "    commonalities = commonProteins(humanProteins, elephantProteins)\n",
    "    differences = findAminoAcidDifferences(humanProteins, elephantProteins)\n",
    "    displayTextResults(commonalities, differences)\n",
    "    Labels = makeAminoAcidLabels([humanProteins, elephantProteins])\n",
    "    freqList = setupChartData(Labels, [humanProteins, elephantProteins])\n",
    "    freqLabels = [\"human\", \"elephant\"]\n",
    "    edge_List = makeEdgeList(Labels, differences)\n",
    "    Chart = createChart(Labels, freqList, freqLabels, edge_List)\n",
    "    return    \n",
    "\n",
    "runFullProgram()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
